src/pygama/raw/buffer_processor/buffer_processor.py

```diff
@@ -0,0 +1,345 @@
+) -> lgdo.WaveformTable:
+    r"""
+    Finds the presum rate that was used to perform the DSP and writes this to a table.
+    The name of a valid :func:`pygama.lgdo.WaveformTable` that DSP was performed on is used to extract the original
+    ``dt`` so that they can be updated by the presum rate. Then the presummed table is added to the rb.lgdo
+
+    Parameters
+    ----------
+    rb
+        A :module:`.RawBuffer` containing an rb.lgdo to store the presummed object in
+    presum_obj
+        The :module:`pygama.lgdo.ArrayofEqualSizedArrays` containing the presummed output from the DSP
+    dsp_dict
+        The dictionary that was used to perform the DSP
+    proc
+        The processor name that called :func:`pygama.dsp._processors.presum` in the DSP
+    """
+    # find the presum rate from the dsp_dict
+    presum_rate_string = dsp_dict["processors"][proc]["args"][1]
+    presum_rate_start_idx = presum_rate_string.find("/") + 1
+    presum_rate_end_idx = presum_rate_string.find(",")
+    presum_rate = int(presum_rate_string[presum_rate_start_idx:presum_rate_end_idx])
+
+    # Find the original lgdo field that was used to create the presummed output
+    rb_lgdo_field = dsp_dict["processors"][proc]["args"][0]
+
+    # make sure that this field is a waveform table, then process dts
+    if isinstance(rb.lgdo[rb_lgdo_field], lgdo.WaveformTable):
+        dt = process_presum_dt(rb.lgdo[rb_lgdo_field]["dt"], presum_rate)
+        t0 = rb.lgdo[rb_lgdo_field]["t0"]
+    else:
+        raise TypeError(f"field {rb_lgdo_field} is not a valid lgdo.WaveformTable")
+
+    # Create the new waveform table
+    new_obj = lgdo.WaveformTable(t0=t0, dt=dt, values=presum_obj.nda)
+
+    # Write the presum_rate as an array to the table as a new field
+    presum_rate_array = lgdo.Array(shape=len(t0), dtype=np.uint16, fill_val=presum_rate)
+    rb.lgdo.add_field("presum_rate", presum_rate_array)
+
+    return new_obj
+
+
+def process_presum_dt(dts: Array, presum_rate: int) -> Array:
+    """
+    Multiply a waveform's `dts` by the presumming rate, used for presummed waveforms.
+    """
+    # don't want to modify the original lgdo_table dts
+    copy_dts = copy.deepcopy(dts)
+
+    # change the dt by the presum rate
+    copy_dts.nda *= presum_rate
+    return copy_dts
+
+
+def process_windowed_t0(t0s: Array, dts: Array, start_index: int) -> Array:
+    """
+    In order for the processed data to work well with :module:`pygama.dsp.build_dsp`, we need
+    to keep ``t0`` in its original units.
+
+    So we transform ``start_index`` to the units of ``t0`` and add it to every
+    ``t0`` value.
+    """
+    # don't want to modify the original lgdo_table dts
+    copy_dts = copy.deepcopy(dts)
+    copy_t0s = copy.deepcopy(t0s)
+
+    # perform t0+start_index*dt to rewrite the new t0 in terms of sample
+    start_index *= copy_dts.nda
+    copy_t0s.nda += start_index
+    return copy_t0s
+
+
+def process_saturation(n_sat: Array) -> None:
```

Comment:
This can be removed since this functionality is handled by the dtype conversion operation.